engines/e_padlock.c

```diff
@@ -452,11 +560,108 @@ padlock_ctr_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out_arg,
+    
+    memcpy(cdata->iv, EVP_CIPHER_CTX_iv(ctx), SM4_BLOCK_SIZE);
+    gmi_sm4_encrypt(out_arg, in_arg, cdata, nbytes);
+    memcpy(EVP_CIPHER_CTX_iv_noconst(ctx), cdata->iv, SM4_BLOCK_SIZE);   
+    return 1;
+}
+
+static int
+gmi_sm4_ctr_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out_arg,
+                   const unsigned char *in_arg, size_t nbytes)
+{
+    struct gmi_cipher_data *cdata = ALIGNED_CIPHER_DATA_GMI(ctx);
+    unsigned int num = EVP_CIPHER_CTX_num(ctx);
+    
+    if (f_zxc == 1) {
+        unsigned char * buf = EVP_CIPHER_CTX_buf_noconst(ctx);
+        CRYPTO_ctr128_encrypt(in_arg, out_arg, nbytes,
+                              cdata->ks.rd_key, EVP_CIPHER_CTX_iv_noconst(ctx), buf, &num,
+                              (block128_f) gmi_sm4_ecb_enc);
+       EVP_CIPHER_CTX_set_num(ctx, num);
+    } else {
+        memcpy(cdata->iv, EVP_CIPHER_CTX_iv(ctx), SM4_BLOCK_SIZE);
+        if (nbytes % SM4_BLOCK_SIZE) {        
+            nbytes = SM4_BLOCK_SIZE - nbytes % SM4_BLOCK_SIZE + nbytes;    
+        }
+        gmi_sm4_encrypt(out_arg, in_arg, cdata, nbytes);
+        memcpy(EVP_CIPHER_CTX_iv_noconst(ctx), cdata->iv, SM4_BLOCK_SIZE);
+        EVP_CIPHER_CTX_set_num(ctx, num);
+    }
+    return 1;
+}
+
+static int
+gmi_sm4_cfb128_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out_arg,
+                      const unsigned char *in_arg, size_t nbytes)
+{
+    struct gmi_cipher_data *cdata = ALIGNED_CIPHER_DATA_GMI(ctx);
+    
+    if (f_zxc == 1) {
+        int num = EVP_CIPHER_CTX_num(ctx);
+        CRYPTO_cfb128_encrypt(in_arg, out_arg, nbytes, cdata->ks.rd_key,
+                              EVP_CIPHER_CTX_iv_noconst(ctx), &num, EVP_CIPHER_CTX_encrypting(ctx),
+                              (block128_f)gmi_sm4_ecb_enc);
+    } else {
+        memcpy(cdata->iv, EVP_CIPHER_CTX_iv(ctx), SM4_BLOCK_SIZE);
+        if (nbytes % SM4_BLOCK_SIZE) {        
+            nbytes = SM4_BLOCK_SIZE - nbytes % SM4_BLOCK_SIZE + nbytes;    
+        }
+        gmi_sm4_encrypt(out_arg, in_arg, cdata, nbytes);
+        memcpy(EVP_CIPHER_CTX_iv_noconst(ctx), cdata->iv, SM4_BLOCK_SIZE);
+    }
+    return 1;
+}
+
+static int
+gmi_sm4_ofb128_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out_arg,
+                      const unsigned char *in_arg, size_t nbytes)
+{
+    struct gmi_cipher_data *cdata = ALIGNED_CIPHER_DATA_GMI(ctx);
+    
+    if (f_zxc == 1) {
+        int num = EVP_CIPHER_CTX_num(ctx);
+        CRYPTO_ofb128_encrypt(in_arg, out_arg, nbytes, cdata->ks.rd_key,
+                              EVP_CIPHER_CTX_iv_noconst(ctx), &num, 
+                              (block128_f)gmi_sm4_ecb_enc);
+    } else {
+        memcpy(cdata->iv, EVP_CIPHER_CTX_iv(ctx), SM4_BLOCK_SIZE);
+        if (nbytes % SM4_BLOCK_SIZE) {        
+            nbytes = SM4_BLOCK_SIZE - nbytes % SM4_BLOCK_SIZE + nbytes;    
+        }
+        gmi_sm4_encrypt(out_arg, in_arg, cdata, nbytes);
+        memcpy(EVP_CIPHER_CTX_iv_noconst(ctx), cdata->iv, SM4_BLOCK_SIZE);
```

Comment:
also here, the number of bytes in the partial block must be updated, (unless I am completely mistaken)